src/compiler/rust.rs

```diff
@@ -1003,18 +1113,72 @@ impl Compilation for RustCompilation {
                     let input_path = Path::new(input_path).to_owned();
                     dist_arguments.push(try_string_arg!(input_path.into_arg_string(path_transformer_fn)))
                 } else {
+                    if let Some(Target(_)) = argument.get_data() {
+                        saw_target = true
+                    }
                     for string_arg in argument.iter_strings(path_transformer_fn) {
                         dist_arguments.push(try_string_arg!(string_arg))
                     }
                 }
             }
 
+            // We can't rely on the packaged toolchain necessarily having the same default target triple
+            // as us (typically host triple), so make sure to always explicitly specify a target.
+            if !saw_target {
+                dist_arguments.push(format!("--target={}", host))
+            }
+
+            // Convert the paths of some important environment variables
+            let mut env_vars = dist::osstring_tuples_to_strings(env_vars)?;
+            let mut changed_out_dir: Option<PathBuf> = None;
+            for (k, v) in env_vars.iter_mut() {
+                match k.as_str() {
+                    // We round-tripped from path to string and back to path, but it should be lossless
+                    "OUT_DIR" => {
+                        let dist_out_dir = path_transformer.to_dist(Path::new(v))?;
+                        if dist_out_dir != *v {
+                            changed_out_dir = Some(v.to_owned().into());
+                        }
+                        *v = dist_out_dir
+                    }
+                    "CARGO" |
+                    "CARGO_MANIFEST_DIR" => {
+                        *v = path_transformer.to_dist(Path::new(v))?
+                    },
+                    _ => (),
+                }
+            }
+            // OUT_DIR was changed during transformation, check if this compilation is relying on anything
+            // inside it - if so, disallow distributed compilation (there are sometimes hardcoded paths present)
```

Comment:
This would happen for things like "my build script compiled some C code" or "my build script generated Rust code", right? Is it too hard to ship those additional files across the wire in those cases, or are there other cases I'm missing?